agent/taskresource/volume/dockervolume_efs.go

```diff
@@ -0,0 +1,128 @@
+
+	// Enums used by acs's api-2.json model.
+	efsIAMAuthEnabled           = "ENABLED"
+	efsTransitEncryptionEnabled = "ENABLED"
+)
+
+// EFSVolumeConfig represents efs volume configuration.
+type EFSVolumeConfig struct {
+	AuthConfig            EFSAuthConfig `json:"authorizationConfig"`
+	FileSystemID          string        `json:"fileSystemId"`
+	RootDirectory         string        `json:"rootDirectory"`
+	TransitEncryption     string        `json:"transitEncryption"`
+	TransitEncryptionPort int64         `json:"transitEncryptionPort"`
+	// DockerVolumeName is internal docker name for this volume.
+	DockerVolumeName string `json:"dockerVolumeName"`
+}
+
+// EFSAuthConfig contains auth config for an efs volume.
+type EFSAuthConfig struct {
+	AccessPointId string `json:"accessPointId"`
+	Iam           string `json:"iam"`
+}
+
+// GetDriverOptions returns the driver options for creating an EFS volume.
+func GetDriverOptions(cfg *config.Config, efsVolCfg *EFSVolumeConfig, credsRelativeURI string) map[string]string {
+	if UseECSVolumePlugin(cfg) {
+		return efsVolCfg.getVolumePluginDriverOptions(credsRelativeURI)
+	}
+	return efsVolCfg.getDockerLocalDriverOptions(cfg)
+}
+
+// UseECSVolumePlugin checks whether we should use the ECS volume plugin.
+func UseECSVolumePlugin(cfg *config.Config) bool {
+	for _, cap := range cfg.VolumePluginCapabilities {
+		if cap == efsVolumePluginCapability {
+			return true
+		}
+	}
+	return false
+}
+
+// GetDockerLocalDriverOptions returns the options for creating an EFS volume using Docker's local volume driver.
+func (efsVolCfg *EFSVolumeConfig) getDockerLocalDriverOptions(cfg *config.Config) map[string]string {
+	domain := getDomainForPartition(cfg.AWSRegion)
+	// These are the NFS options recommended by EFS, see:
+	// https://docs.aws.amazon.com/efs/latest/ug/mounting-fs-mount-cmd-general.html
+	ostr := fmt.Sprintf("addr=%s.efs.%s.%s,nfsvers=4.1,rsize=1048576,wsize=1048576,hard,timeo=600,retrans=2,noresvport", efsVolCfg.FileSystemID, cfg.AWSRegion, domain)
+	devstr := fmt.Sprintf(":%s", efsVolCfg.RootDirectory)
+	return map[string]string{
+		"type":   efsLocalDriverType,
+		"device": devstr,
+		"o":      ostr,
+	}
+}
+
+// GetEFSDriverOptions returns the options for creating an EFS volume using the ECS volume plugin.
+func (efsVolCfg *EFSVolumeConfig) getVolumePluginDriverOptions(credsRelativeURI string) map[string]string {
+	device := efsVolCfg.FileSystemID
+	if efsVolCfg.RootDirectory != "" {
+		device = fmt.Sprintf("%s:%s", device, efsVolCfg.RootDirectory)
+	}
+
+	mntOpt := NewVolumeMountOptions()
+	if efsVolCfg.TransitEncryption == efsTransitEncryptionEnabled {
+		mntOpt.AddOption("tls", "")
+	}
+	if efsVolCfg.TransitEncryptionPort != 0 {
+		mntOpt.AddOption("tlsport", strconv.Itoa(int(efsVolCfg.TransitEncryptionPort)))
+	}
+	if efsVolCfg.AuthConfig.Iam == efsIAMAuthEnabled {
+		mntOpt.AddOption("iam", "")
+		mntOpt.AddOption("awscredentialsrelativeuri", credsRelativeURI)
```

Comment:
isn't `awscredentialsrelativeuri` changed to a different name now?

Reply:
yeah, updated to awscredsuri